service/controller/v13/resource/instance/create_scale_workers.go

```diff
@@ -0,0 +1,80 @@
+package instance
+
+import (
+	"context"
+	"fmt"
+
+	providerv1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/azure-operator/service/controller/v13/key"
+	"github.com/giantswarm/azure-operator/service/controller/v13/resource/instance/internal/state"
+	"github.com/giantswarm/microerror"
+)
+
+func (r *Resource) scaleUpWorkerVMSSTransition(ctx context.Context, obj interface{}, currentState state.State) (state.State, error) {
+	customObject, err := key.ToCustomObject(obj)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	desiredWorkerCount := key.WorkerCount(customObject) * 2
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("scaling worker VMSS to %d nodes", desiredWorkerCount))
+
+	// Double the desired number of nodes in worker VMSS in order to
+	// provide 1:1 mapping between new up-to-date nodes when draining and
+	// terminating old nodes.
+	err = r.scaleVMSS(ctx, customObject, key.WorkerVMSSName, desiredWorkerCount)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("scaled worker VMSS to %d nodes", desiredWorkerCount))
+
+	return CordonOldWorkers, nil
+}
+
+func (r *Resource) scaleDownWorkerVMSSTransition(ctx context.Context, obj interface{}, currentState state.State) (state.State, error) {
+	customObject, err := key.ToCustomObject(obj)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	desiredWorkerCount := key.WorkerCount(customObject)
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("scaling worker VMSS to %d nodes", desiredWorkerCount))
+
+	// Scale down to the desired number of nodes in worker VMSS.
+	err = r.scaleVMSS(ctx, customObject, key.WorkerVMSSName, desiredWorkerCount)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("scaled worker VMSS to %d nodes", desiredWorkerCount))
+
+	return DeploymentCompleted, nil
+}
+
+func (r *Resource) scaleVMSS(ctx context.Context, customObject providerv1alpha1.AzureConfig, deploymentNameFunc func(customObject providerv1alpha1.AzureConfig) string, nodeCount int) error {
+	c, err := r.getScaleSetsClient(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	vmss, err := c.Get(ctx, key.ResourceGroupName(customObject), deploymentNameFunc(customObject))
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	*vmss.Sku.Capacity = int64(nodeCount)
+	res, err := c.CreateOrUpdate(ctx, key.ResourceGroupName(customObject), deploymentNameFunc(customObject), vmss)
```

Comment:
Do we know if Azure has some kind of [optimistic concurrency control](https://en.wikipedia.org/wiki/Optimistic_concurrency_control)? I don't see any other way of doing this than _getting_ and _updating_ the VMSS, but it sucks.

Reply:
Not that I know of. I'm not sure if that would change much in this case. If external party changes VMSS size during update, it might do it even outside of this small scope and results would still be unknown.